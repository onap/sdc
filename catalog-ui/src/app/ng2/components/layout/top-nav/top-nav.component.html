<!--
  ~ Copyright (C) 2018 AT&T Intellectual Property. All rights reserved.
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~      http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
  -->
 
 
<nav class="top-nav">

    <div class="asdc-app-title-wrapper">
        <a class="asdc-app-title">{{ 'PROJECT_TITLE'|translate }}</a>
        <div class="asdc-version"> v.{{version}}</div>
    </div>

    <ul class="top-menu" *ngIf="!menuModel && topLvlMenu">
        <!-- no hierarchy & dropdowns mode -->
        <li *ngFor="let item of topLvlMenu.menuItems; let i = index"
            [ngClass]="{'selected': i == topLvlMenu.selectedIndex}">
            <a (click)="menuItemClick(topLvlMenu, item)"
               [attr.data-tests-id]="'main-menu-button-' + item.text.toLowerCase()">{{item.text}}</a>
        </li>
    </ul>

    <ul class="top-menu" *ngIf="menuModel">
        <!-- with hierarchy & dropdowns mode -->
        <ng-container *ngFor="let groupItem of menuModel; let $index = index; let $last = last">
            <li [ngClass]="{'selected': $last }">
                <a (click)="menuItemClick(groupItem, groupItem.menuItems[groupItem.selectedIndex])"
                   [attr.data-tests-id]="'breadcrumbs-button-' + $index">
                    {{groupItem.menuItems[groupItem.selectedIndex]?.text}}
                </a>
            </li>
            <li class="triangle-dropdown"
                [ngClass]="{'item-click': groupItem.itemClick}" (mouseover)="groupItem.itemClick = true">
                <div class="triangle"><span class="sprite-new arrow-right"></span></div>
                <ul class="sub-menu">
                    <li *ngFor="let ddItem of groupItem.menuItems; let $index2 = index"
                        (click)="menuItemClick(groupItem, ddItem)"
                        [ngClass]="{'selected': $index2 == groupItem.selectedIndex, 'disabled': ddItem.isDisabled}"
                        [attr.data-tests-id]="'sub-menu-button-' + ddItem.text.toLowerCase()">
                        <span sdc-smart-tooltip="">{{ddItem.text}}</span>
                    </li>
                </ul>
            </li>
        </ng-container>
    </ul>

    <div class="top-search" [hidden]="hideSearch === true">
        <input type="text"
               class="search-text"
               placeholder="Search"
               [ngModel]="searchTerm"
               (ngModelChange)="emitSearchTerm($event)"
               data-tests-id="main-menu-input-search" />
        <span class="w-sdc-search-icon magnification"></span>
    </div>

    <div class="notification-icon" [ngClass]="{'disabled' : progress > 0}" *ngIf="user && user.role === 'DESIGNER' && notificationIconCallback" (click)="notificationIconCallback()" tooltip="Vendor Software Product Repository" tooltipPlacement="left" data-tests-id="repository-icon"></div>

</nav>
