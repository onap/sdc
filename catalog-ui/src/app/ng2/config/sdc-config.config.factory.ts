declare const __ENV__:string;

export interface ISdcConfig {
    [index:string]: any
}

export function getSdcConfig() : ISdcConfig {
    let sdcConfig:ISdcConfig = {};

    if (__ENV__==='prod') {
        sdcConfig = require('./../../../../configurations/prod.js');
    } else {
        sdcConfig = require('./../../../../configurations/dev.js');
    }

    return sdcConfig;
}
