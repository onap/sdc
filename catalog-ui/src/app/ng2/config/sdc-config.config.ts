import {Provider, OpaqueToken} from "@angular/core";
import {getSdcConfig, ISdcConfig} from "./sdc-config.config.factory";

export { ISdcConfig };

export const SdcConfigToken = new OpaqueToken('SdcConfigToken');

export const SdcConfig:Provider = {
    provide: SdcConfigToken,
    useFactory: getSdcConfig
};
