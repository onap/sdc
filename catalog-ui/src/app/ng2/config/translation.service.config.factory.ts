import { ITranslateServiceConfig } from "../shared/translator/translate.service.config";

declare const __ENV__:string;

export function getTranslationServiceConfig() : ITranslateServiceConfig {
    const pathPrefix = (__ENV__ === 'prod') ? 'sdc1/' : '';
    return {
        filePrefix: pathPrefix + 'assets/languages/',
        fileSuffix: '.json',
        allowedLanguages: ['en_US'],
        defaultLanguage: 'en_US'
    };
}
