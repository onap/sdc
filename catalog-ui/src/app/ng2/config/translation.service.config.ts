import { Provider } from "@angular/core";
import { TranslateServiceConfigToken } from "../shared/translator/translate.service.config";
import { getTranslationServiceConfig } from "./translation.service.config.factory";

export const TranslationServiceConfig:Provider = {
    provide: TranslateServiceConfigToken,
    useFactory: getTranslationServiceConfig
};
