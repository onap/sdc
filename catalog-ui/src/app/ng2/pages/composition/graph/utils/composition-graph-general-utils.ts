/*-
 * ============LICENSE_START=======================================================
 * SDC
 * ================================================================================
 * Copyright (C) 2017 AT&T Intellectual Property. All rights reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END=========================================================
 */

import * as _ from "lodash";
import {ComponentInstance, Match, CompositionCiLinkBase, CompositionCiNodeUcpeCp} from "app/models";
import {Dictionary, GraphUIObjects} from "app/utils";
import {MatchCapabilitiesRequirementsUtils} from "./match-capability-requierment-utils";
import {CommonGraphUtils} from "../common/common-graph-utils";
import {Injectable} from "@angular/core";
import {QueueServiceUtils} from "app/ng2/utils/queue-service-utils";
import {ComponentServiceNg2} from "app/ng2/services/component-services/component.service";
import {RequirementsGroup} from "app/models/requirement";
import {CapabilitiesGroup} from "app/models/capability";
import {TopologyTemplateService} from "app/ng2/services/component-services/topology-template.service";
import {CompositionService} from "../../composition.service";
import {WorkspaceService} from "app/ng2/pages/workspace/workspace.service";
import {NotificationsService} from "onap-ui-angular/dist/notifications/services/notifications.service";
import {NotificationSettings} from "onap-ui-angular/dist/notifications/utilities/notification.config";

export interface RequirementAndCapabilities {
    capabilities: CapabilitiesGroup;
    requirements: RequirementsGroup;
}

@Injectable()
export class CompositionGraphGeneralUtils {

    public componentRequirementsAndCapabilitiesCaching = new Dictionary<string, RequirementAndCapabilities>();

    constructor(private commonGraphUtils: CommonGraphUtils,
                private matchCapabilitiesRequirementsUtils: MatchCapabilitiesRequirementsUtils,
                private queueServiceUtils: QueueServiceUtils,
                private componentService: ComponentServiceNg2,
                private topologyTemplateService: TopologyTemplateService,
                private compositionService: CompositionService,
                private workspaceService: WorkspaceService) {
    }

    /**
     * Get the offset for the link creation Menu
     * @param point
     * @returns {Cy.Position}
     */
    public calcMenuOffset: Function = (point: Cy.Position): Cy.Position => {
        point.x = point.x + 60;
        point.y = point.y + 105;
        return point;
    };

    /**
     *  return the top left position of the link menu
     * @param cy
     * @param targetNodePosition
     * @returns {Cy.Position}
     */
    public getLinkMenuPosition = (cy: Cy.Instance, targetNodePosition: Cy.Position) => {
        let menuPosition: Cy.Position = this.calcMenuOffset(targetNodePosition); //get the link mid point
        if ($(document.body).height() < menuPosition.y + GraphUIObjects.LINK_MENU_HEIGHT + $(document.getElementsByClassName('sdc-composition-graph-wrapper')).offset().top) { // if position menu is overflow bottom
            menuPosition.y = $(document.body).height() - GraphUIObjects.TOP_HEADER_HEIGHT - GraphUIObjects.LINK_MENU_HEIGHT;
        }
        return menuPosition;
    };

    public zoomGraphTo = (cy: Cy.Instance, zoomLevel: number): void => {
        let zy = cy.height() / 2;
        let zx = cy.width() / 2;
        cy.zoom({
            level: zoomLevel,
            renderedPosition: {x: zx, y: zy}
        });
    }

    //saves the current zoom, and then sets a temporary maximum zoom for zoomAll, and then reverts to old value
    public zoomAllWithMax = (cy: Cy.Instance, maxZoom: number): void => {

        let oldMaxZoom: number = cy.maxZoom();

        cy.maxZoom(maxZoom);
        this.zoomAll(cy);
        cy.maxZoom(oldMaxZoom);

    };

    //Zooms to fit all of the nodes in the collection passed in. If no nodes are passed in, will zoom to fit all nodes on graph
    public zoomAll = (cy: Cy.Instance, nodes?: Cy.CollectionNodes): void => {

        if (!nodes || !nodes.length) {
            nodes = cy.nodes();
        }

        cy.resize();
        cy.animate({
            fit: {eles: nodes, padding: 20},
            center: {eles: nodes}
        }, {duration: 400});
    };

    /**
     * will return true/false if two nodes overlapping
     *
     * @param graph node
     */
    private isNodesOverlapping(node: Cy.CollectionFirstNode, draggedNode: Cy.CollectionFirstNode): boolean {

        let nodeBoundingBox: Cy.BoundingBox = node.renderedBoundingBox();
        let secondNodeBoundingBox: Cy.BoundingBox = draggedNode.renderedBoundingBox();

        return this.isBBoxOverlapping(nodeBoundingBox, secondNodeBoundingBox);
    }

    /**
     * Checks whether the bounding boxes of two nodes are overlapping on any side
     * @param nodeOneBBox
     * @param nodeTwoBBox
     * @returns {boolean}
     */
    private isBBoxOverlapping(nodeOneBBox: Cy.BoundingBox, nodeTwoBBox: Cy.BoundingBox) {
        return (((nodeOneBBox.x1 < nodeTwoBBox.x1 && nodeOneBBox.x2 > nodeTwoBBox.x1) ||
            (nodeOneBBox.x1 < nodeTwoBBox.x2 && nodeOneBBox.x2 > nodeTwoBBox.x2) ||
            (nodeTwoBBox.x1 < nodeOneBBox.x1 && nodeTwoBBox.x2 > nodeOneBBox.x2)) &&
            ((nodeOneBBox.y1 < nodeTwoBBox.y1 && nodeOneBBox.y2 > nodeTwoBBox.y1) ||
                (nodeOneBBox.y1 < nodeTwoBBox.y2 && nodeOneBBox.y2 > nodeTwoBBox.y2) ||
                (nodeTwoBBox.y1 < nodeOneBBox.y1 && nodeTwoBBox.y2 > nodeOneBBox.y2)))
    }

    /**
     * Checks whether a specific topologyTemplate instance can be hosted on the UCPE instance
     * @param cy - Cytoscape instance
     * @param fromUcpeInstance
     * @param toComponentInstance
     * @returns {Match}
     */
    public canBeHostedOn(cy: Cy.Instance, fromUcpeInstance: ComponentInstance, toComponentInstance: ComponentInstance): Match {

        let matches: Array<Match> = this.matchCapabilitiesRequirementsUtils.getMatchedRequirementsCapabilities(fromUcpeInstance, toComponentInstance, this.getAllCompositionCiLinks(cy));
        let hostedOnMatch: Match = _.find(matches, (match: Match) => {
            return match.requirement.capability.toLowerCase() === 'tosca.capabilities.container';
        });

        return hostedOnMatch;
    };

    /**
     * Checks whether node can be dropped into UCPE
     * @param cy
     * @param nodeToInsert
     * @param ucpeNode
     * @returns {boolean}
     */
    private isValidDropInsideUCPE(cy: Cy.Instance, nodeToInsert: ComponentInstance, ucpeNode: ComponentInstance): boolean {

        let hostedOnMatch: Match = this.canBeHostedOn(cy, ucpeNode, nodeToInsert);
        let result: boolean = !angular.isUndefined(hostedOnMatch) || nodeToInsert.isVl(); //group validation
        return result;

    };

    /**
     * For drops from palette, checks whether the node can be dropped. If node is being held over another node, check if capable of hosting
     * @param cy
     * @param pseudoNodeBBox
     * @param paletteComponentInstance
     * @returns {boolean}
     */
    public isPaletteDropValid(cy: Cy.Instance, pseudoNodeBBox: Cy.BoundingBox) {

        let illegalOverlappingNodes = _.filter(cy.nodes("[isSdcElement]"), (graphNode: Cy.CollectionFirstNode) => {
            if (this.isBBoxOverlapping(pseudoNodeBBox, graphNode.renderedBoundingBox())) {
                return true;
            }
            return false;
        });

        return illegalOverlappingNodes.length === 0;
    }

    /**
     * will return true/false if a drop of a single node is valid
     *
     * @param graph node
     */
    public isValidDrop(cy: Cy.Instance, draggedNode: Cy.CollectionFirstNode): boolean {

        let illegalOverlappingNodes = _.filter(cy.nodes("[isSdcElement]"), (graphNode: Cy.CollectionFirstNode) => { //all sdc nodes, removing child nodes (childe node allways collaps

            if (draggedNode.data().isUcpe && (graphNode.isChild() || graphNode.data().isInsideGroup)) { //ucpe cps always inside ucpe, no overlapping
                return false;
            }
            if (draggedNode.data().isInsideGroup && (!draggedNode.active() || graphNode.data().isUcpe)) {
                return false;
            }

            if (!draggedNode.data().isUcpe && !(draggedNode.data() instanceof CompositionCiNodeUcpeCp) && graphNode.data().isUcpe) { //case we are dragging a node into UCPE
                let isEntirelyInUCPE: boolean = this.commonGraphUtils.isFirstBoxContainsInSecondBox(draggedNode.renderedBoundingBox(), graphNode.renderedBoundingBox());
                if (isEntirelyInUCPE) {
                    if (this.isValidDropInsideUCPE(cy, draggedNode.data().componentInstance, graphNode.data().componentInstance)) { //if this is valid insert into ucpe, we return false - no illegal overlapping nodes
                        return false;
                    }
                }
            }
            return graphNode.data().id !== draggedNode.data().id && this.isNodesOverlapping(draggedNode, graphNode);

        });
        // return false;
        return illegalOverlappingNodes.length === 0;
    };

    /**
     * will return true/false if the move of the nodes is valid (no node overlapping and verifying if insert into UCPE is valid)
     *
     * @param  nodesArray - the selected drags nodes
     */
    public isGroupValidDrop(cy: Cy.Instance, nodesArray: Cy.CollectionNodes): boolean {
        let filterDraggedNodes = nodesArray.filter('[?isDraggable]');
        let isValidDrop = _.every(filterDraggedNodes, (node: Cy.CollectionFirstNode) => {
            return this.isValidDrop(cy, node);

        });
        return isValidDrop;
    };

    /**
     * get all links in diagram
     * @param cy
     * @returns {any[]|boolean[]}
     */
    public getAllCompositionCiLinks = (cy: Cy.Instance): Array<CompositionCiLinkBase> => {
        return _.map(cy.edges("[isSdcElement]"), (edge: Cy.CollectionEdges) => {
            return edge.data();
        });
    };

    /**
     *
     * @param blockAction - true/false if this is a block action
     * @param instances
     * @param component
     */
    public pushMultipleUpdateComponentInstancesRequestToQueue = (instances: Array<ComponentInstance>): void => {
        this.queueServiceUtils.addNonBlockingUIAction(() => {
            return new Promise<boolean>((resolve, reject) => {
                let uniqueId = this.workspaceService.metadata.uniqueId;
                let topologyType = this.workspaceService.metadata.componentType;
                this.topologyTemplateService.updateMultipleComponentInstances(uniqueId, topologyType, instances).subscribe(instancesResult => {
                    this.compositionService.updateComponentInstances(instancesResult);
                    resolve(true);
                });
            });
        });
    }
}
