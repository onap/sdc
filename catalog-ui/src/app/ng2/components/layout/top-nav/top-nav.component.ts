/*-
 * ============LICENSE_START=======================================================
 * SDC
 * ================================================================================
 * Copyright (C) 2017 AT&T Intellectual Property. All rights reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END=========================================================
 */

import * as _ from "lodash";
import {Component, Inject, Input, Output, EventEmitter, OnInit, OnDestroy, OnChanges} from "@angular/core";
import {IHostedApplication, IUserProperties} from "app/models";
import {MenuItemGroup, MenuItem} from "app/utils";
import {AuthenticationService} from "../../../services/authentication.service";
import {SdcConfigToken, ISdcConfig} from "../../../config/sdc-config.config";
import {TranslateService} from "../../../shared/translator/translate.service";
import {PluginsConfiguration, Plugin} from "app/models";
import { Subscription } from "rxjs";
// import { Store } from "@ngrx/store";
// import { AppState } from "app/ng2/store/app.state";
// import * as unsavedChangesReducer from 'app/ng2/store/reducers/unsaved-changes.reducer';

declare const window:any;
@Component({
    selector: 'top-nav',
    templateUrl: './top-nav.component.html',
    styleUrls:['./top-nav.component.less']
})
export class TopNavComponent implements OnInit, OnChanges {
    @Input() public version:string;
    @Input() public menuModel:Array<MenuItemGroup>;
    @Input() public topLvlSelectedIndex:number;
    @Input() public hideSearch:boolean;
    @Input() public searchTerm:string;
    @Input() public notificationIconCallback:Function;
    @Input() public unsavedChanges: boolean;
    @Input() public unsavedChangesCallback: (completeCallback:Function)=> Promise<any>;
    @Output() public searchTermChange:EventEmitter<string> = new EventEmitter<string>();
    emitSearchTerm(event:string) {
        this.searchTermChange.emit(event);
    }

    private subscription: Subscription;
    private hasUnsavedChanges: boolean;
    public topLvlMenu:MenuItemGroup;
    public user:IUserProperties;
    private topNavPlugins: Array<Plugin>;

    constructor(private translateService:TranslateService,
                @Inject('$state') private $state:ng.ui.IStateService,
                private authService:AuthenticationService,
                @Inject(SdcConfigToken) private sdcConfig:ISdcConfig) {
        window.nav = this;

    }

    private _getTopLvlSelectedIndexByState = ():number => {
        if (!this.topLvlMenu.menuItems) {
            return 0;
        }

        let result = -1;

        //set result to current state
        this.topLvlMenu.menuItems.every((item:MenuItem, index:number)=> {
            if (item.state === this.$state.current.name) {
                if (this.$state.current.name === 'plugins') {
                    const pluginIdx = _.findIndex(this.topNavPlugins, (plugin: Plugin) => plugin.pluginStateUrl === this.$state.params.path);
                    if (pluginIdx !== -1) {
                        result = index + pluginIdx;
                        return false;
                    }
                } else {
                    result = index;
                    return false;
                }
            }
            return true;
        });

        //if it's a different state
        if (result === -1) {
            //if in 'workspace' -  checking previous state param
            if (this.$state.includes('workspace')) {
                // if previous state is 'dashboard' or 'catalog', then select it - otherwise, use 'catalog' as default for 'workspace'
                const selectedStateName = (['dashboard', 'catalog'].indexOf(this.$state.params['previousState']) !== -1)
                    ? this.$state.params['previousState']
                    : 'catalog';
                result = this.topLvlMenu.menuItems.findIndex((item:MenuItem) => item.state === selectedStateName);
            }

            //if yet, none is selected, then select the first as default
            if (result === -1) {
                result = 0;
            }
        }

        return result;
    };

    ngOnChanges(changes) {
        if (changes['menuModel']) {
            console.log('menuModel was changed!');
            this.generateMenu();
        }
    }

    ngOnInit() {
        console.log('Nav is init!', this.menuModel);
        this.user = this.authService.getLoggedinUser();
        this.topNavPlugins = _.filter(PluginsConfiguration.plugins, (plugin: Plugin) => {
            return plugin.pluginDisplayOptions["tab"] !== undefined;
        });

        this.translateService.languageChangedObservable.subscribe((lang) => {
            let tmpArray: Array<MenuItem> = [
                new MenuItem(this.translateService.translate("TOP_MENU_HOME_BUTTON"), null, "dashboard", "goToState", null, null),
                new MenuItem(this.translateService.translate("TOP_MENU_CATALOG_BUTTON"), null, "catalog", "goToState", null, null)
            ];

            // Only designer can perform onboarding
            if (this.user && this.user.role === 'DESIGNER') {
                tmpArray.push(new MenuItem(this.translateService.translate("TOP_MENU_ON_BOARD_BUTTON"), null, "onboardVendor", "goToState", null, null));
                _.each(this.sdcConfig.hostedApplications, (hostedApp: IHostedApplication) => {
                    if (hostedApp.exists) {
                        tmpArray.push(new MenuItem(hostedApp.navTitle, null, hostedApp.defaultState, "goToState", null, null));
                    }
                });
            }

            // Adding plugins to top-nav only if they can be displayed for the current connected user role
            _.each(PluginsConfiguration.plugins, (plugin: Plugin) => {
                if (plugin.pluginDisplayOptions["tab"] && (this.user && plugin.pluginDisplayOptions["tab"].displayRoles.includes(this.user.role))) {
                    tmpArray.push(new MenuItem(plugin.pluginDisplayOptions["tab"].displayName, null, "plugins", "goToState", {path: plugin.pluginStateUrl}, null));
                }
            });

            this.topLvlMenu = new MenuItemGroup(0, tmpArray, true);
            this.topLvlMenu.selectedIndex = isNaN(this.topLvlSelectedIndex) ? this._getTopLvlSelectedIndexByState() : this.topLvlSelectedIndex;
            this.generateMenu();
        });
    }

    generateMenu() {
        if (this.menuModel && this.topLvlMenu && this.menuModel[0] !== this.topLvlMenu) {
            this.menuModel.unshift(this.topLvlMenu);
        }
    }

    goToState(state:string, params:any):Promise<boolean> {
        return new Promise((resolve, reject) => {
            this.$state.go(state, params || undefined);
            resolve(true);
        });
    }


    menuItemClick(itemGroup:MenuItemGroup, item:MenuItem) {

        let onSuccessFunction = () => {
            this.navigate(itemGroup, item);
        }
        if (this.unsavedChanges && this.unsavedChangesCallback){
            this.unsavedChangesCallback(onSuccessFunction).then((onSuccess)=> {
                this.navigate(itemGroup, item);
            }, (onReject) => {});
        } else {
            this.navigate(itemGroup, item);
        }
    }

    navigate(itemGroup:MenuItemGroup, item:MenuItem) {
        itemGroup.itemClick = false;
        let onSuccess = ():void => {
            itemGroup.selectedIndex = itemGroup.menuItems.indexOf(item);
        };
        let onFailed = ():void => {
        };

        if (item.callback) {
            (item.callback.apply(undefined, item.params)).then(onSuccess, onFailed);
        } else {
            this[item.action](item.state, item.params).then(onSuccess, onFailed);
        }
    }

}
