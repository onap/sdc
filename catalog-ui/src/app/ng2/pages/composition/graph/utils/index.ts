/**
 * Created by ob0695 on 6/3/2018.
 */
// export * from './composition-graph-general-utils';
// export * from './composition-graph-links-utils';
// export * from './composition-graph-nodes-utils';
// export * from './composition-graph-palette-utils';
// export * from './composition-graph-service-path-utils';
// export * from './composition-graph-zone-utils';


import {CompositionGraphGeneralUtils} from './composition-graph-general-utils';
import {CompositionGraphNodesUtils} from './composition-graph-nodes-utils';
import {MatchCapabilitiesRequirementsUtils} from './match-capability-requirement-utils'
import {CompositionGraphPaletteUtils} from './composition-graph-palette-utils';
import {CompositionGraphZoneUtils} from './composition-graph-zone-utils';
import {ServicePathGraphUtils} from './composition-graph-service-path-utils';
import {CompositionGraphLinkUtils} from "./composition-graph-links-utils";


export {
    CompositionGraphGeneralUtils,
    CompositionGraphLinkUtils,
    CompositionGraphNodesUtils,
    MatchCapabilitiesRequirementsUtils,
    CompositionGraphPaletteUtils,
    CompositionGraphZoneUtils,
    ServicePathGraphUtils
};