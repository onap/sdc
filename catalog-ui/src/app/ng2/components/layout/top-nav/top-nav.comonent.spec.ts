import { NO_ERRORS_SCHEMA } from '@angular/core';
import { async, ComponentFixture } from '@angular/core/testing';
import { PluginsConfiguration } from 'app/models';
import { Observable } from 'rxjs';
import { Mock } from 'ts-mockery';
import { ConfigureFn, configureTests } from '../../../../../jest/test-config.helper';
import { MenuItem, MenuItemGroup } from '../../../../utils/menu-handler';
import { SdcConfigToken } from '../../../config/sdc-config.config';
import { AuthenticationService } from '../../../services/authentication.service';
import { TranslateModule } from '../../../shared/translator/translate.module';
import { TranslateService } from '../../../shared/translator/translate.service';
import { TopNavComponent } from './top-nav.component';

describe('artifact form component', () => {

    let fixture: ComponentFixture<TopNavComponent>;
    let translateServiceMock: Partial<TranslateService>;
    let mockStateService;
    let authServiceMock;

    const designerUser = {
        email: 'designer@sdc.com',
        firstName: 'Carlos',
        fullName: 'Carlos Santana',
        lastLoginTime: 1555587266566,
        lastName: 'Santana',
        role: 'DESIGNER',
        status: 'ACTIVE',
        userId: 'cs0008'
    };

    const pluginDisplayOptions = {
        displayName : '',
        displayContext : new Array<string>(),
        displayRoles :  new Array<string>()
    };

    let roleToReturn = designerUser;

    const map1 =  new Map();
    map1.otherValue =  pluginDisplayOptions;

    const map2 = new Map();
    pluginDisplayOptions.displayRoles = ['DESIGNER'];
    pluginDisplayOptions.displayName = 'DCAE-DS';
    map2.tab =  pluginDisplayOptions;

    PluginsConfiguration.plugins =
        [
            {pluginId: 'DCAED', pluginDiscoveryUrl: 'DCAED_discoveryURL', pluginSourceUrl: 'DCAED_sourceURL', pluginStateUrl: 'DCAED_stateURL', pluginDisplayOptions:  map1, isOnline: true},
            {pluginId: 'DCAE-DS', pluginDiscoveryUrl: 'DCAE-DS_discoveryURL', pluginSourceUrl: 'DCAE-DS_sourceURL', pluginStateUrl: 'DCAE-DS_stateURL', pluginDisplayOptions: map2, isOnline: true}
        ];

    beforeEach(
        async(() => {
            authServiceMock = {
                getLoggedinUser: jest.fn().mockImplementation(() => {
                    return roleToReturn;
                })
            };

            mockStateService = {
                go: jest.fn(),
                includes: jest.fn(() => true),
                current : {
                    name : 'plugins'
                },
                params : {}
            };

            translateServiceMock = {
                languageChangedObservable: Mock.of<Observable<string>>( {
                    subscribe : jest.fn().mockImplementation((cb) => {
                        cb();
                    })
                }),
                translate: jest.fn((str: string) => {
                    if (str === 'TOP_MENU_HOME_BUTTON') {
                        return 'HOME';
                    } else if (str === 'TOP_MENU_CATALOG_BUTTON') {
                        return 'CATALOG';
                    } else if (str === 'TOP_MENU_ON_BOARD_BUTTON') {
                        return 'ONBOARD';
                    } else {
                        return 'TBD...';
                    }
                })
            };

            const configure: ConfigureFn = (testBed) => {
                testBed.configureTestingModule({
                    declarations: [TopNavComponent],
                    imports: [TranslateModule],
                    schemas: [NO_ERRORS_SCHEMA],
                    providers: [
                        {provide: TranslateService, useValue: translateServiceMock},
                        {provide: '$state', useValue: mockStateService},
                        {provide: AuthenticationService, useValue: authServiceMock},
                        {provide: SdcConfigToken, useValue: {csarFileExtension: 'csar', toscaFileExtension: 'yaml,yml'}},
                    ],
                });
            };

            configureTests(configure).then((testBed) => {
                fixture = testBed.createComponent(TopNavComponent);
            });
        })
    );

    it('should match current snapshot of top-nav component', () => {
        expect(fixture).toMatchSnapshot();
    });

    it('Once a Designer logged in, Menu Items will contain HOME, CATALOG, ONBOARD & DCAE-DS; HOME will be selected', () => {

        // topLvlSelectedIndex = 1 => Ignore the inner call to  _getTopLvlSelectedIndexByState.
        fixture.componentInstance.topLvlSelectedIndex = 0;
        fixture.componentInstance.ngOnInit();

        expect(fixture.componentInstance.topLvlMenu.itemClick).toBe(true);

        expect(fixture.componentInstance.topLvlMenu.menuItems.length).toBe(4);

        expect(fixture.componentInstance.topLvlMenu.menuItems[0]).toEqual({
            action: 'goToState',
            blockedForTypes: null,
            callback: null,
            params: null,
            state: 'dashboard',
            text: 'HOME'
        });
        expect(fixture.componentInstance.topLvlMenu.menuItems[1]).toEqual({
            action: 'goToState',
            blockedForTypes: null,
            callback: null,
            params: null,
            state: 'catalog',
            text: 'CATALOG'
        });
        expect(fixture.componentInstance.topLvlMenu.menuItems[2]).toEqual({
            action: 'goToState',
            blockedForTypes: null,
            callback: null,
            params: null,
            state: 'onboardVendor',
            text: 'ONBOARD'
        });
        expect(fixture.componentInstance.topLvlMenu.menuItems[3]).toEqual({
            action: 'goToState',
            blockedForTypes: null,
            callback: null,
            params:
                {path: 'DCAE-DS_stateURL'},
            state: 'plugins',
            text: 'DCAE-DS'
        });

        expect(fixture.componentInstance.topLvlMenu.selectedIndex).toBe(0);
    });

});
