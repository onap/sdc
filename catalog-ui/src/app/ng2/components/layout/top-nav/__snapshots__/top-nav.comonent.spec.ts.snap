// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`artifact form component should match current snapshot of top-nav component 1`] = `
<top-nav
  $state={[Function Object]}
  _getTopLvlSelectedIndexByState={[Function Function]}
  authService={[Function Object]}
  sdcConfig={[Function Object]}
  searchTermChange={[Function EventEmitter]}
  translateService={[Function Object]}
>
  <nav
    class="top-nav"
  >
    <div
      class="asdc-app-title-wrapper"
    >
      <a
        class="asdc-app-title"
      >
        
      </a>
      <div
        class="asdc-version"
      >
         v.
      </div>
    </div>
    
    
    <div
      class="top-search"
    >
      <input
        class="search-text"
        data-tests-id="main-menu-input-search"
        placeholder="Search"
        type="text"
      />
      <span
        class="w-sdc-search-icon magnification"
      />
    </div>
    
  </nav>
</top-nav>
`;
